cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -11,3 +11,119 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 package main
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	corev1 "github.com/kubeapps/kubeapps/cmd/kubeapps-apis/gen/core/packages/v1alpha1"
+	datapackagingv1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apiserver/apis/datapackaging/v1alpha1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	log "k8s.io/klog/v2"
+)
+
+// GetAvailablePackageSummaries returns the available packages managed by the 'kapp_controller' plugin
+func (s *Server) GetAvailablePackageSummaries(ctx context.Context, request *corev1.GetAvailablePackageSummariesRequest) (*corev1.GetAvailablePackageSummariesResponse, error) {
+	log.Infof("+kapp-controller GetAvailablePackageSummaries")
+
+	// Retrieve the proper parameters from the request
+	namespace := request.GetContext().GetNamespace()
+	cluster := request.GetContext().GetCluster()
+	pageSize := request.GetPaginationOptions().GetPageSize()
+	pageOffset, err := pageOffsetFromPageToken(request.GetPaginationOptions().GetPageToken())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to intepret page token %q: %v", request.GetPaginationOptions().GetPageToken(), err)
+	}
+	// Assume the default cluster if none is specified
+	if cluster == "" {
+		cluster = s.globalPackagingCluster
+	}
+	// fetch all the package metadatas
+	pkgMetadatas, err := s.getPkgMetadatas(ctx, cluster, namespace)
+	if err != nil {
+		return nil, errorByStatus("get", "PackageMetadata", "", err)
+	}
+
+	// paginate the list of results
+	availablePackageSummaries := make([]*corev1.AvailablePackageSummary, len(pkgMetadatas))
+
+	// create the waiting group for processing each item aynchronously
+	var wg sync.WaitGroup
+
+	if len(pkgMetadatas) > 0 {
+		startAt := -1
+		if pageSize > 0 {
+			startAt = int(pageSize) * pageOffset
+		}
+		for i, pkgMetadata := range pkgMetadatas {
+			wg.Add(1)
+			if startAt <= i {
+				go func(i int, pkgMetadata *datapackagingv1alpha1.PackageMetadata) error {
```

Comment:
I think this is the same/similar method you're using elsewhere? (sorry, it was hours ago in a different review, not 100%), in which case, perhaps we can DRY it up (eventually, no problem if it's just a comment for now).

Reply:
Yep, I've used this pagination logic (inspired by Greg's code :P) also in the GetInstalledPackageSummaries. Adding a comment in both places for future improvement. Thanks!